verifiable-db/src/revelation/revelation_unproven_offset.rs

```diff
@@ -0,0 +1,1295 @@
+                                                       // computed by the universal query circuit
+                                                       // closure to access the output items of the i-th result
+        let get_result = |i| &results[S * i..S * (i + 1)];
+        let [min_query, max_query] = b.add_virtual_u256_arr_unsafe(); // unsafe should be ok since they are later included in placeholder hash
+        let [limit, offset] = b.add_virtual_target_arr();
+        let tree_hash = original_tree_proof.merkle_hash();
+        let zero = b.zero();
+        let one = b.one();
+        let zero_u256 = b.zero_u256();
+        let _true = b._true();
+        let _false = b._false();
+        let mut num_results = zero;
+        let placeholder_hash = row_proofs[0].placeholder_hash_target();
+        let computational_hash = row_proofs[0].computational_hash_target();
+        let mut overflow = _false;
+        let mut row_paths = vec![];
+        let mut index_paths = vec![];
+        let mut max_result = None;
+        // Flag employed to enforce that the matching rows are all placed in the initial slots;
+        // this is a requirement to ensure that the check for DISTINCT is sound
+        let mut only_matching_rows = _true;
+        row_proofs
+            .into_iter()
+            .enumerate()
+            .for_each(|(i, row_proof)| {
+                let index_ids = row_proof.index_ids_target();
+                let is_matching_row = b.is_equal(row_proof.num_matching_rows_target(), one);
+                // ensure that once `is_matching_row = false`, then it will be false for all
+                // subsequent iterations
+                only_matching_rows = b.and(only_matching_rows, is_matching_row);
+                b.connect(only_matching_rows.target, is_matching_row.target);
+                let row_node_hash = {
+                    // if the node storing the current row is a leaf node in rows tree, then
+                    // the hash of such node is already computed by `row_proof`; otherwise,
+                    // we need to compute it
+                    let inputs = row_node_info[i]
+                        .child_hashes
+                        .into_iter()
+                        .flat_map(|hash| hash.to_targets())
+                        .chain(row_node_info[i].node_min.to_targets())
+                        .chain(row_node_info[i].node_max.to_targets())
+                        .chain(once(index_ids[1]))
+                        .chain(row_proof.min_value_target().to_targets())
+                        .chain(row_proof.tree_hash_target().to_targets())
+                        .collect_vec();
+                    let row_node_hash = b.hash_n_to_hash_no_pad::<H>(inputs);
+                    b.select_hash(
+                        is_row_node_leaf[i],
+                        &row_proof.tree_hash_target(),
+                        &row_node_hash,
+                    )
```

Comment:
why enforcing that select if we anyway compute it and we have the data ? Can't we just use the one computed ?

Reply:
Because we need to use the hash fetched from the row proof, generated with the universal circuit. As an optimization for the other type of queries, this circuit will already compute the hash of the node if the row is stored in a leaf node, otherwise the hash exposed as public input will be the hash of the embedded cells tree. So, when processing this proof, we need to enforce that we are recomputing the hash of the node correctly from the hash exposed as public input by the row proof